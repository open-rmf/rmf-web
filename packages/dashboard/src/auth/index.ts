export * from './authenticator';
export * from './keycloak';
export * from './login-card';
export * from './login-page';
export * from './private-route';
export * from './stub';
export * from './user-profile';
