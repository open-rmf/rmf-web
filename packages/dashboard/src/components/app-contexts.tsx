import React from 'react';
import ResourceManager from '../resource-manager';
import { defaultSettings, Settings } from '../settings';
import { NotificationBarContext, NotificationBarProps } from './notification-bar';

/* Declares the ResourcesContext which contains the resources used on the app*/
export const ResourcesContext = React.createContext<ResourceManager | undefined>(undefined);

export const SettingsContext = React.createContext(defaultSettings());

export const TooltipContext = React.createContext({
  showTooltips: true,
  toggleTooltips: () => {},
});

export interface AppContextProviderProps extends React.PropsWithChildren<{}> {
  settings: Settings;
  resourceManager?: ResourceManager;
  notificationDispatch: React.Dispatch<React.SetStateAction<NotificationBarProps | null>>;
  tooltips: { showTooltips: boolean; toggleTooltips: () => void };
}

export function AppContextProvider(props: AppContextProviderProps): React.ReactElement {
  const { settings, notificationDispatch, resourceManager, tooltips, children } = props;
  return (
    <SettingsContext.Provider value={settings}>
      <NotificationBarContext.Provider value={notificationDispatch}>
        <TooltipContext.Provider value={tooltips}>
          <ResourcesContext.Provider value={resourceManager}>{children}</ResourcesContext.Provider>
        </TooltipContext.Provider>
      </NotificationBarContext.Provider>
    </SettingsContext.Provider>
  );
}
