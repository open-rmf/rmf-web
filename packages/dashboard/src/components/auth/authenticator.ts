import Debug from 'debug';
import EventEmitter from 'eventemitter3';
import Keycloak_, { KeycloakInstance } from 'keycloak-js';
import { BASE_PATH } from '../../util/url';
import { User } from './user';

const debug = Debug('authenticator');

export type AuthenticatorEventType = {
  userChanged: [User | null];
};

export default interface Authenticator extends EventEmitter<AuthenticatorEventType> {
  readonly user?: User;
  readonly token?: string;

  /**
   * It is required to call this before using any of the authenticator functions.
   */
  init(): Promise<void>;

  /**
   * Note: This redirects to external login page so it will never return.
   */
  login(): Promise<never>;

  /**
   * Note: This redirects to external logout page so it will never return.
   */
  logout(): Promise<void>;
}

export type AuthConfig<T = typeof Keycloak_> = T extends (config: infer U) => unknown ? U : never;

export class DefaultAuthenticator
  extends EventEmitter<AuthenticatorEventType>
  implements Authenticator {
  get user(): User | undefined {
    return this._user;
  }

  get token(): string | undefined {
    return this._inst.idToken;
  }

  constructor(config: AuthConfig, redirectUri?: string) {
    super();
    this._inst = Keycloak_(config);
    this._redirectUri = redirectUri;
  }

  async init() {
    debug('initializing authenticator');

    this._inst.onAuthSuccess = async () => {
      this._user = {
        username: (this._inst.idTokenParsed as any).preferred_username,
      };
      debug('authenticated as', this._user.username);
      this.emit('userChanged', this._user);
    };

    this._inst.onAuthLogout = () => {
      debug('logout');
      this._user = undefined;
      this.emit('userChanged', null);
    };

    await this._inst.init({
      onLoad: 'check-sso',
      silentCheckSsoRedirectUri: `${BASE_PATH}/silent-check-sso.html`,
      redirectUri: this._redirectUri,
    });
    try {
      const refreshed = await this._inst.updateToken(30);
      refreshed && debug('token refreshed');
    } catch {}

    this._user = this._inst.idTokenParsed && {
      username: (this._inst.idTokenParsed as any).preferred_username,
    };
  }

  async login(): Promise<never> {
    await this._inst.login();
    throw new Error('should not reach here');
  }

  async logout(): Promise<never> {
    await this._inst.logout();
    throw new Error('should not reach here');
  }

  private _inst: KeycloakInstance;
  private _redirectUri?: string;
  private _user?: User;
}
