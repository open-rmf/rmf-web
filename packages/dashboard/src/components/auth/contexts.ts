import { Permission, User as ApiUser } from 'api-client';
import React from 'react';

export type UserProfile = ApiUser;

export interface User {
  profile: UserProfile;
  permissions: Permission[];
}

export const UserContext = React.createContext<User | null>(null);
