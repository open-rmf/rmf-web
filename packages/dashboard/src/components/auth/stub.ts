import EventEmitter from 'eventemitter3';
import Authenticator, { AuthenticatorEventType } from './authenticator';
import { User } from './user';

export default class StubAuthenticator
  extends EventEmitter<AuthenticatorEventType>
  implements Authenticator {
  readonly user: User = {
    username: 'stub',
  };

  readonly token?: string = undefined;

  init(): Promise<void> {
    return Promise.resolve();
  }

  login(): Promise<never> {
    throw new Error('not supported');
  }

  logout(): Promise<never> {
    throw new Error('not supported');
  }
}
