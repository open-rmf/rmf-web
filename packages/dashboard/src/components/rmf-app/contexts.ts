import { Dispenser, Fleet, Ingestor } from 'api-client';
import React from 'react';
import * as RmfModels from 'rmf-models';
import { HealthStatus } from '../../managers/rmf-health-state-manager';
import { Place } from './place';
import { RmfIngress } from './rmf-ingress';

export const BuildingMapContext = React.createContext<RmfModels.BuildingMap | null>(null);
export const PlacesContext = React.createContext<Place[]>([]);
export const FleetsContext = React.createContext<Fleet[]>([]);
export const DispensersContext = React.createContext<Dispenser[]>([]);
export const IngestorsContext = React.createContext<Ingestor[]>([]);

const itemSummary = () => {
  return {
    operational: 0,
    spoiltItem: [],
  };
};

const initializeItemSummary = itemSummary();

export const RmfHealthContext = React.createContext<HealthStatus>({
  door: { ...initializeItemSummary },
  lift: { ...initializeItemSummary },
  dispenser: { ...initializeItemSummary },
  robot: {
    operational: 0,
    idle: 0,
    charging: 0,
    spoiltRobots: [],
  },
});

export const RmfIngressContext = React.createContext<RmfIngress | undefined>(undefined);
