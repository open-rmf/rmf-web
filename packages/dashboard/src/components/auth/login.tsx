import { Button } from '@material-ui/core';
import React from 'react';
import { LoginPage } from 'react-components';
import { Redirect } from 'react-router';
import { DASHBOARD_ROUTE, getUrl } from '../../util/url';
import { AuthenticatorContext, UserContext } from '../auth/contexts';

export interface LoginProps {
  /**
   * defaults to url of `DASHBOARD_ROUTE`
   */
  successRedirectUri?: string;
}

export default function Login(props: LoginProps): JSX.Element {
  const successRedirectUri = props.successRedirectUri || getUrl(DASHBOARD_ROUTE);
  const authenticator = React.useContext(AuthenticatorContext);
  const user = React.useContext(UserContext);

  async function handleRmfLogin(_event: React.MouseEvent): Promise<void> {
    authenticator.login(successRedirectUri);
  }

  return user ? (
    <Redirect to={successRedirectUri} />
  ) : (
    <LoginPage title="RoMi Dashboard" logo="assets/ros-health.png">
      <Button id="login-button" onClick={handleRmfLogin} variant="contained">
        Login with RMF
      </Button>
    </LoginPage>
  );
}
