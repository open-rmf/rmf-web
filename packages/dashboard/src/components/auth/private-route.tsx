import Debug from 'debug';
import React from 'react';
import { Redirect, Route, RouteProps, useLocation } from 'react-router';
import { LOGIN_ROUTE } from '../../util/url';
import Unauthorized from '../error-pages/unauthorized';
import { UserContext } from './contexts';

const debug = Debug('PrivateRoute');

export interface PrivateRouteProps extends React.PropsWithChildren<RouteProps> {
  // if true, do not redirect to login url if not authenticated
  noRedirectToLogin?: boolean;
}

/**
 * This component validates if the user is authenticated before rendering component passed as a
 * prop.
 */
const PrivateRoute = ({
  noRedirectToLogin,
  children,
  ...rest
}: PrivateRouteProps): JSX.Element | null => {
  const user = React.useContext(UserContext);
  const location = useLocation();

  function render() {
    if (user) {
      return children;
    } else {
      if (!noRedirectToLogin) {
        debug('accessing private route while unauthenticated');
        debug('redirecting to login page');
        return <Redirect to={{ pathname: LOGIN_ROUTE, state: { from: location } }} />;
      } else {
        return <Unauthorized />;
      }
    }
  }

  return <Route {...rest}>{render()}</Route>;
};

export default PrivateRoute;
