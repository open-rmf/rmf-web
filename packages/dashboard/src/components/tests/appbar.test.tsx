import { render } from '@testing-library/react';
import { act } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { AppController, AppControllerContext } from '../app-contexts';
import AppBar from '../appbar';
import { AuthenticatorContext, User, UserContext } from '../auth/contexts';
import FakeAuthenticator from '../auth/__mocks__/fake-authenticator';
import { makeMockAppController } from './mock-app-controller';
import { mountAsUser } from './test-utils';

describe('AppBar', () => {
  let appController: AppController;
  const Base = (props: React.PropsWithChildren<{}>) => {
    return (
      <AppControllerContext.Provider value={appController}>
        {props.children}
      </AppControllerContext.Provider>
    );
  };

  beforeEach(() => {
    appController = makeMockAppController();
  });

  test('renders with navigation bar', () => {
    const root = render(
      <Base>
        <AppBar tabValue="building" />
      </Base>,
    );
    expect(root.getAllByRole('tablist').length > 0).toBeTruthy();
  });

  test('show settings when settings button is clicked', () => {
    const root = render(
      <Base>
        <AppBar tabValue="building" />
      </Base>,
    );
    act(() => {
      const elements = root.getAllByTestId('setting-tooltip-tooltip');
      for (let element in elements) {
        userEvent.click(elements[element]);
      }
    });
    expect(appController.showSettings).toBeCalledTimes(1);
  });

  test('shows help when help button is clicked', () => {
    const root = render(
      <Base>
        <AppBar tabValue="building" />
      </Base>,
    );
    act(() => {
      const elements = root.getAllByTestId('help-tooltip-tooltip');
      for (let element in elements) {
        userEvent.click(elements[element]);
      }
    });
    expect(appController.showHelp).toBeCalledTimes(1);
  });

  test('renders tooltips when it is enabled', async () => {
    const root = render(
      <Base>
        <AppBar tabValue="building" />
      </Base>,
    );
    userEvent.hover(root.getByTestId('help-tooltip-tooltip'));
    expect(await root.findByText('Help tools and resources')).toBeTruthy();

    userEvent.hover(root.getByTestId('setting-tooltip-tooltip'));
    expect(await root.findByText('Define dashboard trajectory settings')).toBeTruthy();
  });

  test('user button is shown when there is an authenticated user', () => {
    const user: User = {
      profile: { username: 'test', is_admin: false, roles: [] },
      permissions: [],
    };
    const root = mountAsUser(
      user,
      <Base>
        <AppBar tabValue="building" />
      </Base>,
    );
    expect(root.getByLabelText('user-btn')).toBeTruthy();
  });

  test('logout is triggered when logout button is clicked', () => {
    const authenticator = new FakeAuthenticator('test');
    const spy = jest.spyOn(authenticator, 'logout').mockImplementation(() => undefined as any);
    const user: User = {
      profile: { username: 'test', is_admin: false, roles: [] },
      permissions: [],
    };
    const root = render(
      <Base>
        <AuthenticatorContext.Provider value={authenticator}>
          <UserContext.Provider value={user}>
            <AppBar tabValue="building" />
          </UserContext.Provider>
        </AuthenticatorContext.Provider>
      </Base>,
    );
    userEvent.click(root.getByLabelText('user-btn'));
    userEvent.click(root.getByText('Logout'));
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
