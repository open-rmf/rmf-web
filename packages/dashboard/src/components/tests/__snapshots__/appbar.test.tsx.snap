// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar renders correctly 1`] = `
<WithStyles(ForwardRef(AppBar))
  id="appbar"
  position="static"
>
  <WithStyles(ForwardRef(Toolbar))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-toolbarTitle-1"
      variant="h6"
    >
      Dashboard
    </WithStyles(ForwardRef(Typography))>
    <DashboardTooltip
      enabled={true}
      id="omnipanel-tooltip"
      title="View all available panel options"
    >
      <WithStyles(ForwardRef(IconButton))
        color="inherit"
        id="toggle-omnipanel-btn"
        onClick={[Function]}
      >
        <DashboardIcon />
      </WithStyles(ForwardRef(IconButton))>
    </DashboardTooltip>
    <DashboardTooltip
      enabled={true}
      id="setting-tooltip"
      title="Define dashboard trajectory settings"
    >
      <WithStyles(ForwardRef(IconButton))
        color="inherit"
        id="show-settings-btn"
        onClick={[Function]}
      >
        <SettingsIcon />
      </WithStyles(ForwardRef(IconButton))>
    </DashboardTooltip>
    <DashboardTooltip
      enabled={true}
      id="help-tooltip"
      title="Help tools and resources"
    >
      <WithStyles(ForwardRef(IconButton))
        color="inherit"
        id="show-help-btn"
        onClick={[Function]}
      >
        <HelpIcon />
      </WithStyles(ForwardRef(IconButton))>
    </DashboardTooltip>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
